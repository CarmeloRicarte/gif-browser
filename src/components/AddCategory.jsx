import { useState } from "react";
import PropTypes from "prop-types";
export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  /**
   * When the input changes, set the input value to the value of the input.
   */
  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  /**
   * The function onSubmit calls the function onNewCategory, which is passed in as a prop from the
   * parent component.
   * The function onNewCategory is passed the value of the input field.
   * @returns The input value is being returned.
   */
  const onSubmit = (event) => {
    event.preventDefault();
    const newInputValue = inputValue.trim();
    if (newInputValue.length <= 1) {
      return;
    }
    onNewCategory(newInputValue);
    setInputValue("");
  };

  return (
    <form aria-label="form" onSubmit={onSubmit}>
      <input
        type="text"
        id="seachGifs"
        placeholder="Buscar gifs"
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
