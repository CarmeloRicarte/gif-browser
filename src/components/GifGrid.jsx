import { useEffect, useState } from "react";
import { getGifs } from "./../helpers/getGifs";
import { GifItem } from "./GifItem";

export const GifGrid = ({ category }) => {
  const [images, setImages] = useState([]);

  /**
   * GetImages is an async function that calls getGifs, which is an async function that returns a
   * promise, and then sets the state of images to the value of the promise
   */
  const getImages = async () => {
    const newImages = await getGifs(category);
    setImages(newImages);
  };

  useEffect(() => {
    getImages();
  }, []);

  return (
    <>
      <h3>{category}</h3>

      <div>
        {images.map((image) => (
          <GifItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};
