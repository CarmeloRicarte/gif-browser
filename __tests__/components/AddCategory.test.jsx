import { fireEvent, render, screen } from "@testing-library/react";
import { AddCategory } from "../../src/components/AddCategory";
describe("Test of <AddCategory />", () => {
  test("should match with snapshot", () => {
    const { container } = render(<AddCategory onNewCategory={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  test("should change the value of input", () => {
    // create test subject
    render(<AddCategory onNewCategory={() => {}} />);
    // extract the input. Only there is one textbox
    const input = screen.getByRole("textbox");

    // simulate onChange event for applying the new value
    fireEvent.change(input, { target: { value: "Ibai" } });
    expect(input.value).toBe("Ibai");
  });

  test("should call onNewCategory if the input has a value", () => {
    const inputValue = "Ibai";
    const onNewCategory = jest.fn(); // mock for onNewCategory

    render(<AddCategory onNewCategory={onNewCategory} />);

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.change(input, { target: { value: inputValue } });
    fireEvent.submit(form);
    expect(input.value).toBe("");
    expect(onNewCategory).toHaveBeenCalledTimes(1);
    // evaluates that onNewCategory is called with the input value
    expect(onNewCategory).toHaveBeenCalledWith(inputValue);
  });

  test("should not call onNewCategory if the input hasn't less or equal one character", () => {
    const inputValue = "a";
    const onNewCategory = jest.fn(); // mock for onNewCategory

    render(<AddCategory onNewCategory={onNewCategory} />);

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.change(input, { target: { value: inputValue } });
    fireEvent.submit(form);
    screen.debug();
    expect(onNewCategory).not.toHaveBeenCalled();
  });
});
