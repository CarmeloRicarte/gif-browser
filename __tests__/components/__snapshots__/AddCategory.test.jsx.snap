// Vitest Snapshot v1

exports[`Test of <AddCategory /> > should match with snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      id="seachGifs"
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
